File: algorithms/python/firstMissingPositive/firstMissingPositive.py
# question : https://leetcode.com/problems/first-missing-positive/
# Author : Samir Rajesh Prajapati
# Date : 2020-10-01

# ********************************************************************************* 
# * 
# * Given an unsorted integer array, find the first missing positive integer.
# * 
# * For example,
# * Given [1,2,0] return 3,
# * and [3,4,-1,1] return 2.
# * 
# * Your algorithm should run in O(n) time and uses constant space.
# * 
# *               
# **********************************************************************************

class Solution:
    def firstMissingPositive(self, nums: [int]) -> int:
        for i in range(len(nums)):
            while nums[i]>0 and nums[i]<len(nums) and nums[i]-1!=i and nums[i]!=nums[nums[i]-1]:
                nums[nums[i]-1], nums[i] = nums[i], nums[nums[i]-1]
        for i in range(len(nums)):
            if i + 1 != nums[i]:
                return i + 1 
        return len(nums) + 1
                
if __name__ == '__main__':
    number = Solution()
    print(number.firstMissingPositive([3,4,-1,1]))

File: algorithms/cpp/first_missing_positive.cpp
// problem : https://leetcode.com/problems/first-missing-positive/

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        long int l = nums.size();
        if(l==0)
            return 1;
        if(l==1)
            return nums[0]==1?2:1;
        long int f=0,i;
        for( i=0;i<l;i++)
        {
            if(nums[i]<=0)
            {
                long int v= nums[i];
                nums[i]=nums[f];
                nums[f]=v;
                f++;
            }
        }
        long int st=f;
        cout<<f<<" ";
        for(int i=f;i<l;i++)
        {
            long int in = abs(nums[i])-1+st;
            cout<<in<<" ";
            if(in<l&&nums[in]>=0)
                nums[in]=-nums[in];
        }
        
        for(int i=st;i<l;i++)
            if(nums[i]>=0)
                return i+1-st;
        return l+1-st;
    }
};
